File: Backend/SchoolManagement.Application/Services/SalaryService.cs
using AutoMapper;
using Microsoft.EntityFrameworkCore;
using SchoolManagement.Application.Interfaces;
using SchoolManagement.Domain.Entities;
using SchoolManagement.Infrastructure.Data;
using SchoolManagement.Shared.DTOs.Common;
using SchoolManagement.Shared.DTOs.Salaries;

namespace SchoolManagement.Application.Services;

public class SalaryService : ISalaryService
{
    private readonly SchoolDbContext _context;
    private readonly IMapper _mapper;

    public SalaryService(SchoolDbContext context, IMapper mapper)
    {
        _context = context;
        _mapper = mapper;
    }

    public async Task<PagedResult<SalaryDto>> GetAllSalariesAsync(int pageNumber, int pageSize)
    {
        var query = _context.Salaries
            .Include(s => s.Teacher)
            .ThenInclude(t => t.User)
            .Where(s => !s.IsDeleted)
            .OrderByDescending(s => s.Year)
            .ThenByDescending(s => s.Month);

        var totalCount = await query.CountAsync();
        var salaries = await query
            .Skip((pageNumber - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        var salaryDtos = salaries.Select(s => new SalaryDto
        {
            Id = s.Id,
            TeacherId = s.TeacherId,
            TeacherName = s.Teacher.User.FullName,
            Month = s.Month,
            Year = s.Year,
            BaseSalary = s.BaseSalary,
            HoursWorked = s.HoursWorked,
            Bonus = s.Bonus,
            Deductions = s.Deductions,
            TotalSalary = s.TotalSalary,
            Status = s.Status,
            PaidDate = s.PaidDate,
            CreatedAt = s.CreatedAt,
            UpdatedAt = s.UpdatedAt
        }).ToList();

        return new PagedResult<SalaryDto>
        {
            Data = salaryDtos,
            TotalCount = totalCount,
            PageNumber = pageNumber,
            PageSize = pageSize
        };
    }

    public async Task<SalaryDto?> GetSalaryByIdAsync(int id)
    {
        var salary = await _context.Salaries
            .Include(s => s.Teacher)
            .ThenInclude(t => t.User)
            .FirstOrDefaultAsync(s => s.Id == id && !s.IsDeleted);

        if (salary == null) return null;

        return new SalaryDto
        {
            Id = salary.Id,
            TeacherId = salary.TeacherId,
            TeacherName = salary.Teacher.User.FullName,
            Month = salary.Month,
            Year = salary.Year,
            BaseSalary = salary.BaseSalary,
            HoursWorked = salary.HoursWorked,
            Bonus = salary.Bonus,
            Deductions = salary.Deductions,
            TotalSalary = salary.TotalSalary,
            Status = salary.Status,
            PaidDate = salary.PaidDate,
            CreatedAt = salary.CreatedAt,
            UpdatedAt = salary.UpdatedAt
        };
    }

    public async Task<PagedResult<SalaryDto>> GetTeacherSalariesAsync(int teacherId, int pageNumber, int pageSize)
    {
        var query = _context.Salaries
            .Include(s => s.Teacher)
            .ThenInclude(t => t.User)
            .Where(s => s.TeacherId == teacherId && !s.IsDeleted)
            .OrderByDescending(s => s.Year)
            .ThenByDescending(s => s.Month);

        var totalCount = await query.CountAsync();
        var salaries = await query
            .Skip((pageNumber - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        var salaryDtos = salaries.Select(s => new SalaryDto
        {
            Id = s.Id,
            TeacherId = s.TeacherId,
            TeacherName = s.Teacher.User.FullName,
            Month = s.Month,
            Year = s.Year,
            BaseSalary = s.BaseSalary,
            HoursWorked = s.HoursWorked,
            Bonus = s.Bonus,
            Deductions = s.Deductions,
            TotalSalary = s.TotalSalary,
            Status = s.Status,
            PaidDate = s.PaidDate,
            CreatedAt = s.CreatedAt,
            UpdatedAt = s.UpdatedAt
        }).ToList();

        return new PagedResult<SalaryDto>
        {
            Data = salaryDtos,
            TotalCount = totalCount,
            PageNumber = pageNumber,
            PageSize = pageSize
        };
    }

    public async Task<SalaryDto> CreateSalaryAsync(CreateSalaryDto createSalaryDto)
    {
        var salary = new Salary
        {
            TeacherId = createSalaryDto.TeacherId,
            Month = createSalaryDto.Month,
            Year = createSalaryDto.Year,
            BaseSalary = createSalaryDto.BaseSalary,
            HoursWorked = createSalaryDto.HoursWorked,
            Bonus = createSalaryDto.Bonus,
            Deductions = createSalaryDto.Deductions,
            TotalSalary = createSalaryDto.TotalSalary,
            Status = createSalaryDto.Status,
            PaidDate = createSalaryDto.PaidDate,
            CreatedAt = DateTime.UtcNow
        };

        _context.Salaries.Add(salary);
        await _context.SaveChangesAsync();

        return await GetSalaryByIdAsync(salary.Id) ?? throw new Exception("Failed to create salary");
    }

    public async Task<SalaryDto> UpdateSalaryAsync(int id, CreateSalaryDto updateSalaryDto)
    {
        var salary = await _context.Salaries
            .FirstOrDefaultAsync(s => s.Id == id && !s.IsDeleted);

        if (salary == null)
            throw new Exception("Salary not found");

        salary.TeacherId = updateSalaryDto.TeacherId;
        salary.Month = updateSalaryDto.Month;
        salary.Year = updateSalaryDto.Year;
        salary.BaseSalary = updateSalaryDto.BaseSalary;
        salary.HoursWorked = updateSalaryDto.HoursWorked;
        salary.Bonus = updateSalaryDto.Bonus;
        salary.Deductions = updateSalaryDto.Deductions;
        salary.TotalSalary = updateSalaryDto.TotalSalary;
        salary.Status = updateSalaryDto.Status;
        salary.PaidDate = updateSalaryDto.PaidDate;
        salary.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();

        return await GetSalaryByIdAsync(id) ?? throw new Exception("Failed to update salary");
    }

    public async Task<bool> DeleteSalaryAsync(int id)
    {
        var salary = await _context.Salaries
            .FirstOrDefaultAsync(s => s.Id == id && !s.IsDeleted);

        if (salary == null)
            return false;

        salary.IsDeleted = true;
        salary.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();
        return true;
    }
}

File: Frontend/src/app/core/services/salary.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface SalaryRecord {
  id: number;
  teacherId: number;
  teacherName: string;
  month: number;
  year: number;
  baseSalary: number;
  hoursWorked: number;
  bonus: number;
  deductions: number;
  totalSalary: number;
  status: number;
  paidDate?: Date;
  createdAt: Date;
  updatedAt?: Date;
  allowances: number;
  netSalary: number;
}

export interface CreateSalaryRequest {
  teacherId: number;
  month: number;
  year: number;
  baseSalary: number;
  hoursWorked: number;
  bonus: number;
  deductions: number;
  totalSalary: number;
  status: number;
  paidDate?: Date;
}

export interface PagedResult<T> {
  data: T[];
  items: T[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
  totalPages: number;
  hasPreviousPage: boolean;
  hasNextPage: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class SalaryService {
  private readonly API_URL = `${environment.apiUrl}/salaries`;

  constructor(private http: HttpClient) {}

  getSalaryRecords(pageNumber: number = 1, pageSize: number = 10, teacherId?: number, month?: number, year?: number): Observable<PagedResult<SalaryRecord>> {
    let params = new HttpParams()
      .set('pageNumber', pageNumber.toString())
      .set('pageSize', pageSize.toString());
    
    if (teacherId) {
      params = params.set('teacherId', teacherId.toString());
    }
    
    if (month) {
      params = params.set('month', month.toString());
    }
    
    if (year) {
      params = params.set('year', year.toString());
    }
    
    return this.http.get<PagedResult<SalaryRecord>>(this.API_URL, { params });
  }

  getTeacherSalaries(teacherId: number, pageNumber: number = 1, pageSize: number = 10): Observable<PagedResult<SalaryRecord>> {
    const params = new HttpParams()
      .set('pageNumber', pageNumber.toString())
      .set('pageSize', pageSize.toString());
    
    return this.http.get<PagedResult<SalaryRecord>>(`${this.API_URL}/teacher/${teacherId}`, { params });
  }

  createSalaryRecord(salary: CreateSalaryRequest): Observable<SalaryRecord> {
    return this.http.post<SalaryRecord>(this.API_URL, salary);
  }

  updateSalaryRecord(id: number, salary: CreateSalaryRequest): Observable<SalaryRecord> {
    return this.http.put<SalaryRecord>(`${this.API_URL}/${id}`, salary);
  }

  deleteSalaryRecord(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  paySalary(id: number): Observable<SalaryRecord> {
    return this.http.patch<SalaryRecord>(`${this.API_URL}/${id}/pay`, {});
  }
}